### DISCUSSION TITLE
Disable video and audio and clipboad access

### DISCUSSION THREAD

**Discussion by crippledfaith** [cefsharp/CefSharp#4803] [General]
How to disable webcam and microphone and clipboad access for cefsharp browser.
I tried PermissionHandler  but it does not hit in any of the functions.

I test with https://webrtc.github.io/test-pages/src/audio-and-video/

    public class BrowserPermissionHandler : PermissionHandler
    {
        protected override bool OnShowPermissionPrompt(IWebBrowser chromiumWebBrowser, IBrowser browser, ulong promptId, string requestingOrigin, PermissionRequestType requestedPermissions, IPermissionPromptCallback callback)
        {
            using (callback)
            {
                System.Diagnostics.Debug.WriteLine($"{promptId}|{requestedPermissions} {requestingOrigin}");
                callback.Continue(PermissionRequestResult.Deny);
                return true;
            }
        }

        protected override void OnDismissPermissionPrompt(IWebBrowser chromiumWebBrowser, IBrowser browser, ulong promptId, PermissionRequestResult result)
        {
            base.OnDismissPermissionPrompt(chromiumWebBrowser, browser, promptId, result);
        }

        protected override bool OnRequestMediaAccessPermission(IWebBrowser chromiumWebBrowser, IBrowser browser, IFrame frame, string requestingOrigin, MediaAccessPermissionType requestedPermissions, IMediaAccessCallback callback)
        {
            using (callback)
            {
                //Allow the requested permissions
                //callback.Continue(requestedPermissions);
                return false;
            }
        }
    }

I tried 

    settings.CefCommandLineArgs.Add("enable-media-stream", "0");
    settings.CefCommandLineArgs.Add("enable-experimental-web-platform-features", "0");
    settings.CefCommandLineArgs.Add("enable-usermedia-screen-capturing", "0");

**Comment by amaitland** [comment]
By default media access should be disabled when using the alloy runtime (the default currently). If you've set any of the media access command line arguments then `PermissionHandler` won't be called.

You should be able to use the `disable-javascript-access-clipboard` command line argument to disable clipboard access.

> settings.CefCommandLineArgs.Add("enable-media-stream", "0");

You are enabling media access by using this command line argument, the "0" is always **ignored**. Same would apply to the other arguments you've listed.

If you are using one of the `CefSharp` examples then you will need to remove the command line args that enable media access.

**Comment by crippledfaith** [comment]
Thanks @amaitland i did as you asked but the every time I visit google or the test page (https://webrtc.github.io/test-pages/src/audio-and-video/) kaspersky Internet Security prompts me the CefSharp.BrosererSubprocess is try to access my webcam,
Even thou is does not work in the webpage. how come the CefSharp does want to access my webcam?

![image](https://github.com/cefsharp/CefSharp/assets/7411687/c4d6d82d-19dd-463e-9196-80fee27343ca)

**Reply by amaitland** [reply]
Does `Chrome` supporting disabling of `WebRTC`? Last time I checked it didn't. You can do your own research on this.

> By default media access should be disabled when using the alloy runtime (the default currently).

I should rephrase this.

 `WebRTC` should be blocked by default. The feature should still be enabled.

> kaspersky Internet Security prompts me the CefSharp.BrosererSubprocess is try to access my webcam,

Are you sure the message `kaspersky` displays is actually correct? 

`WebRTC` uses `mDNS` which will trigger a firewall prompt if you haven't added a rule.

Try disabling `mDNS` as per https://github.com/cefsharp/CefSharp/issues/3075#issuecomment-692144916
